### ISSUE TITLE
dynet parameter subset update error

### ISSUE THREAD

**Issue by TheElephantInTheRoom** [clab/dynet#1392]
Take this as an example, it works fine. 
```
import dynet as dy
import numpy as np

model = dy.Model()
pW = model.add_parameters((2, 4))
pb = model.add_parameters(2)
trainer = dy.SimpleSGDTrainer(model)

def step(update_b):

    x = dy.inputTensor(np.ones(4))
    W = pW.expr()
    # update b?
    b = pb.expr(update = update_b)

    loss = dy.pickneglogsoftmax(W * x + b, 0)
    loss.backward()
    trainer.update()
    dy.renew_cg()

print(pb.as_array())
print(pW.as_array())
step(True)
print(pb.as_array()) # b updated
print(pW.as_array())
step(False)
print(pb.as_array()) # b not updated
print(pW.as_array())
```
If you just replace the order of step(False), and step(True) it gives an error: 

>  loss = dy.pickneglogsoftmax(W * x + b, 0)
>   File "_dynet.pyx", line 1899, in _dynet.Expression.__add__
> NotImplementedErro

This could be avoided by moving `dy.renew_cg()` to the beginning of step function. But, still I believe this is a bug.
